src/crimson/os/seastore/transaction.h

```diff
@@ -30,19 +30,32 @@ class Transaction {
   enum class get_extent_ret {
     PRESENT,
     ABSENT,
-    RETIRED
+    RETIRED,
+    INVALIDATED
   };
-  get_extent_ret get_extent(paddr_t addr, CachedExtentRef *out) {
+  get_extent_ret get_extent(paddr_t addr, CachedExtentRef *out, bool delete_on_invalid = false) {
     if (retired_set.count(addr)) {
       return get_extent_ret::RETIRED;
     } else if (auto iter = write_set.find_offset(addr);
 	iter != write_set.end()) {
+      if (!iter->is_valid()) {
+	if (delete_on_invalid) {
+	  write_set.erase(*iter);
+	}
+	return get_extent_ret::INVALIDATED;
+      }
       if (out)
 	*out = CachedExtentRef(&*iter);
       return get_extent_ret::PRESENT;
     } else if (
       auto iter = read_set.find(addr);
       iter != read_set.end()) {
+      if (!(*iter)->is_valid()) {
+	if (delete_on_invalid) {
+	  read_set.erase(*iter);
+	}
+	return get_extent_ret::INVALIDATED;
```

Comment:
I don't want to encode this behavior here.  All this method is meant to do is to reflect whether the extent is present in the transaction so that the caller can then check the larger cache or ultimately the disk.  Yes, an extent in the read_set of a transaction may be invalid, but that's not the only way to encounter one.

Reply:
Also, there's no reason to remove the extent from the read_set or write_set -- the only valid behavior at this point is to propogate an eagain, which must result in a fresh transaction.

Reply:
Probably sufficient to simply check the returned extent in each caller if PRESENT is returned.